### PR TITLE
modemmanager: add watcher for auto reconnect

### PR DESCRIPTION
**Pull request by osedlbauer**
Maintainer: @mips171 / @aleksander0m
Compile tested: x86_64, lantiq_xrx200 latest openwrt master
Run tested: x86_64, lantiq_xrx200 latest openwrt master, test done = yes

Description:
This pr is related to #19794

### CODE REVIEW

**Comment by yegorich** [comment]
Works for me: TI am335x and SIMCom 7600SA-H.

**Comment by yegorich** [comment]
@osedlbauer could you rebase your branch?

**Comment by osedlbauer** [comment]
> @osedlbauer could you rebase your branch?

Done

**Comment by feckert** [comment]
I will make some integration tests tomorrow.

File: net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+main() {
+	local disconnect_detected=0
+
+	mm_log "info" "Start modem state watcher for $MODEM_PATH"
+
+	trap watcher_clean_up INT TERM
+
+	while read -r line; do
+		mm_log "debug" "State changed: $line"
+		if echo "$line" | grep -q "'connected' --> 'disconnecting'"; then
```
Comment (yegorich) [inline net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher:26]: I had some disconnects with another state transition: `state changed (connected -> registered)`. They have to be handled too.

This state transition has been caused by the following reasons:

- mmgdsi-card-event
- implicitly-detached

File: net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+			mm_log "info" "Disconnect detected"
+			disconnect_detected=1
+			break
+		fi
+	done < <(mmcli --monitor-state --modem "$MODEM_PATH")
```
Comment (yegorich) [inline net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher:31]: I am not sure, it is a real error, but my bash syntax checker says:

Syntax error: redirection unexpected

**Comment by yegorich** [comment]
Here some logs for the reference:

```
Wed Jan 24 03:10:44 2024 daemon.warn [2983]: <wrn> [modem0] network reject indication received
Wed Jan 24 03:10:44 2024 daemon.warn [2983]: <wrn> [modem0]   service domain: ps
Wed Jan 24 03:10:44 2024 daemon.warn [2983]: <wrn> [modem0]   radio interface: lte
Wed Jan 24 03:10:44 2024 daemon.warn [2983]: <wrn> [modem0]   reject cause: implicitly-detached
Wed Jan 24 03:10:45 2024 daemon.notice [2983]: <msg> [modem0/bearer7] verbose call end reason (3,1009): [cm] implicitly-detached
Wed Jan 24 03:10:45 2024 daemon.notice [2983]: <msg> [modem0] state changed (connected -> registered)
Wed Jan 24 03:10:45 2024 daemon.notice [2983]: <msg> [modem0/bearer7] connection #1 finished: duration 42740s, tx: 3074314 bytes, rx: 9748794 bytes
Wed Jan 24 03:10:45 2024 daemon.debug ModemManager[6828]: hotplug: State changed: /org/freedesktop/ModemManager1/Modem/0: State changed, 'connected' --> 'registered' (Reason: User request)
Wed Jan 24 03:10:45 2024 daemon.notice [2983]: <msg> [modem0] 3GPP packet service state changed (attached -> detached)
Wed Jan 24 03:10:45 2024 user.notice modemmanager: interface Mobile (network device wwan0) disconnected
Wed Jan 24 03:10:45 2024 daemon.notice netifd: Interface 'Mobile' has lost the connection
Wed Jan 24 03:10:45 2024 daemon.info dnsmasq[4190]: reading /tmp/resolv.conf.d/resolv.conf.auto
Wed Jan 24 03:10:45 2024 daemon.info dnsmasq[4190]: using nameserver 213.209.99.202#53
Wed Jan 24 03:10:45 2024 daemon.info dnsmasq[4190]: using nameserver 192.168.1.1#53
Wed Jan 24 03:10:45 2024 daemon.info dnsmasq[4190]: using only locally-known addresses for test
Wed Jan 24 03:10:45 2024 daemon.notice netifd: Network device 'wwan0' link is down
```

```
Tue Jan 23 14:18:37 2024 daemon.notice [2983]: <msg> [modem0/bearer5] verbose call end reason (2,218): [internal] mmgdsi-card-event
Tue Jan 23 14:18:37 2024 daemon.notice [2983]: <msg> [modem0] state changed (connected -> registered)
Tue Jan 23 14:18:37 2024 daemon.debug ModemManager[5939]: hotplug: State changed: /org/freedesktop/ModemManager1/Modem/0: State changed, 'connected' --> 'registered' (Reason: User request)
Tue Jan 23 14:18:37 2024 daemon.notice [2983]: <msg> [modem0/bearer5] connection #1 finished: duration 10503s, tx: 1241676 bytes, rx: 2051216 bytes
Tue Jan 23 14:18:37 2024 daemon.notice [2983]: <msg> [modem0] 3GPP packet service state changed (attached -> detached)
Tue Jan 23 14:18:37 2024 user.notice modemmanager: interface Mobile (network device wwan0) disconnected
Tue Jan 23 14:18:38 2024 daemon.notice netifd: Interface 'Mobile' has lost the connection
Tue Jan 23 14:18:38 2024 daemon.notice [2983]: <msg> [modem0] 3GPP registration state changed (home -> searching)
Tue Jan 23 14:18:38 2024 daemon.info [2983]: <inf> [modem0] consolidated registration state: cs 'searching', ps 'searching', eps 'searching', 5gs 'unknown' --> 'searching'
Tue Jan 23 14:18:38 2024 daemon.notice [2983]: <msg> [modem0] state changed (registered -> searching)
Tue Jan 23 14:18:38 2024 daemon.info [2983]: <inf> [modem0] access technology changed (lte -> unknown)
Tue Jan 23 14:18:38 2024 daemon.debug ModemManager[5939]: hotplug: State changed: /org/freedesktop/ModemManager1/Modem/0: State changed, 'registered' --> 'searching' (Reason: None or unknown)
```

**Comment by yegorich** [comment]
This hack worked for me:

```
--- a/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher
+++ b/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher
@@ -16,6 +16,8 @@ watcher_clean_up() {

 main() {
        local disconnect_detected=0
+       local state_one="'connected' --> 'disconnecting'"
+       local state_two="'connected' --> 'registered'"

        mm_log "info" "Start modem state watcher for $MODEM_PATH"

@@ -23,7 +25,7 @@ main() {

        while read -r line; do
                mm_log "debug" "State changed: $line"
-               if echo "$line" | grep -q "'connected' --> 'disconnecting'"; then
+               if echo "$line" | grep -q "$state_one\|$state_two"; then
                        mm_log "info" "Disconnect detected"
                        disconnect_detected=1
                        break

```

**Comment by osedlbauer** [comment]
Sorry for the late response, I'm a bit stressed at the moment. The state change seems justified, I can add your changes. However, I'm not sure if process substitution is the right way to go, as it's normally intended for bash and we are using an ash shell. I will take a closer look at this next week.

File: net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher (hunk 1 above)
Comment (aleksander0m) [reply net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher:26]: Every transition out of "connected" is a disconnection

**Comment by yegorich** [comment]
What about this variant?

```diff
diff --git a/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher b/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher
index dfdf2cfba..466d3f446 100755
--- a/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher
+++ b/net/modemmanager/files/usr/libexec/modemmanager/ModemManager-watcher
@@ -23,7 +23,7 @@ main() {

        while read -r line; do
                mm_log "debug" "State changed: $line"
-               if echo "$line" | grep -q "'connected' --> 'disconnecting'"; then
+               if echo "$line" | grep -q "'connected' -->"; then
                        mm_log "info" "Disconnect detected"
                        disconnect_detected=1
                        break
```

**Comment by osedlbauer** [comment]
As I mentioned before, handling this via post-process substitution is problematic since we are not using bash. Therefore, stopping the background processes is a bit tricky. After some consideration with @feckert, we believe the best approach would be to implement an extension of the modem manager, as suggested by @dangowrt (c646b6f34b43619811597f635cc657dbb26ecf33). We are currently waiting for feedback from the upstream ModemManager project. 
Here is a staged pull request: https://github.com/TDT-AG/packages/tree/pr/20240130-modemmanager-check-state

**Comment by aleksander0m** [comment]
> As I mentioned before, handling this via post-process substitution is problematic since we are not using bash. Therefore, stopping the background processes is a bit tricky. After some consideration with @feckert, we believe the best approach would be to implement an extension of the modem manager, as suggested by @dangowrt ([c646b6f](https://github.com/openwrt/packages/commit/c646b6f34b43619811597f635cc657dbb26ecf33)). We are currently waiting for feedback from the upstream ModemManager project. 

A `--track-state` action in mmcli doesn't sound bad to track the state changes of a modem. Maybe named `--monitor-state` instead, to follow the same naming as in the `--monitor-modems` action that already exists.

Could you please send a new merge request to upstream MM with the suggested change?

**Comment by feckert** [comment]
@aleksander0m
The name `--monitor-state` is already in use
```
root@router ~ # mmcli --help-all | grep monitor
  mmcli [OPTION…] - Control and monitor the ModemManager
  -M, --monitor-modems                                           List available modems and monitor additions and removals
  -w, --monitor-state                                            Monitor state of a given modem
  --location-monitor 
```

**Comment by yegorich** [comment]
@dangowrt could create a PR [here](https://gitlab.freedesktop.org/mobile-broadband/ModemManager/-/merge_requests)? All other issues can be resolved there.

**Comment by feckert** [comment]
I will close this. The pull request is replaced by https://github.com/openwrt/packages/pull/23388

**Pull request closed without merging**